src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -108,14 +105,11 @@ private DecisionTree AddByValue(DecisionTree.Guarded guarded, BoundConstantPatte
         {
             if (guarded.Default != null)
             {
-                if (guarded.Default.MatchIsComplete)
-                {
-                    return null;
-                }
+                Debug.Assert(!guarded.Default.MatchIsComplete); // otherwise we would have given a subsumption error
             }
             else
             {
-                guarded.Default = new DecisionTree.ByValue(guarded.Expression, guarded.Type, null);
+                guarded.Default = new DecisionTree.ByType(guarded.Expression, guarded.Type, null);
```

Comment:
It was not correct because the guarded type was not necessarily the type of the constant value. A `ByType` decision tree checks the type first. There are tests that catch this included in this PR.

---
In reply to: [100600126](https://github.com/dotnet/roslyn/pull/16928#discussion_r100600126) [](ancestors = 100600126,100598334,100474813)